lib/foreman/logging.rb

```diff
@@ -103,7 +103,13 @@ def exception(context_message, exception, options = {})
       end
       # backtrace have its own separate level to prevent flooding logs with backtraces
       logger(logger_name).public_send(backtrace_level) do
-        "Backtrace for '#{context_message}' error (#{exception.class}): #{exception.message}\n" + backtrace.join("\n")
+        "Backtrace for '#{context_message}' error (#{exception.class}): #{exception.message}"
+      end
+      # we need a prompt (header) for every backtrace line because it's easier to find by grep
+      backtrace.each do |line|
+        logger(logger_name).public_send(backtrace_level) do
+          line
+        end
```

Comment:
It's easier for grep but on the other hand it's more inconvenient for users of ELK. I suggest if you could leave it as is and take advantage that you have a rake task that should help with finding the relevant data. Can you simply make a code that shows relevant lines when it sees this "Backtrace for" message?

Reply:
The rake task can't find the backlog without this change since the request id isn't logged currently on the backtrace.

Reply:
I don't understand, this is just a normal "debug" logging line, therefore with the default configuration it should have request id automatically prepended alongside with time, date and other things.

Reply:
It doesn't. Because the backtrace is multi line, the metadata is only logged on the first line.

Reply:
Isn't that a problem in our file formatter? Tools like Sentry (and I guess ELK) will hate this because you lose all context and get a wave of exceptions. IMHO this is a bad idea and we should rather fix our file logger if needed.

Reply:
> Because the backtrace is multi line, the metadata is only logged on the first line

Yes, I believe this is the correct behavior. That's why I am suggesting to modify the rake task to actually find those cases and then display all appended lines without prefix. Since we have it, why not to take advantage of this?

Multiline logs are really useful, you have the whole exception in a single log entry when you have integrated logging with something like ElasticSearch. It should be up to the logging formatter how to handle this, I am with @ekohl here. For example syslog logging formatter must break this into individual lines because syslog "defacto standard" does not allow newlines (but most implementations accepts them just fine). We actually have a `MultilinePatternLayout` layout in Foreman which prepends all lines with `|` character to indicate it's continuation, so this can be easily used in the rake task.

In short, let's keep the old behavior and either change the rake task to find those cases or fix the file formatter in logging gem to handle this correctly. However I warn you that the implementation could be pretty complicated in the logging gem, layouts cannot split incoming log messages, it would need probably to be a separate appender implementation. I am not sure if this is worth the effort.

Reply:
I understand that can cause several problems and It's better to stay with current implementation (especially with external log services like the Elastic one) but also with the suggestions from @lzap I'm facing the problem that there is one file and many components that want to log something in async way - like DynFlow. With the suggest of @lzap there is possibly that the multi line log can be splitted into more parts and parsing is much more harder. That's reason why I and @ares bring the adding the "header" to every line. Even the multi line one. 

But I think I have a solution - What if this "header" will be active only if the foreman logging to the file? Rake task is intended to work only file file logging (and @tbrisker take it into account for error page) so this improvement will affect only log-file scenarios. Not the external one.

What do you think, @lzap ?

Reply:
> But I think I have a solution - What if this "header" will be active only if the foreman logging to the file?

I am sorry to say I don't like this. Problem is that a multiline log is actually one operation in logging stack, you need to actually go and split those lines manually. Of course you can do this in Foreman.exception method, but there are several places in the codebase which sends multiple lines via normal logging methods, its not just our codebase but also in our dependencies (this can happen and I'd be surprised if there are none). If you do this, we will end up with mess - some multilines will be formatted with the pattern you propose and some won't.

And you cannot do this in logging gem due to it's design easily. So the only good solution would be override all logging methods and scan all strings for newlines breaking them. I am not fan of that compared to what this brings to us.

Also keep in mind that there are multiple projects parsing our logs trying to find errors or collect important things. This is a breaking change, it will cause some paints to those users.

For these reasons I still think it would be easier and more cleaner to avoid modification of logging in this patch and modify your rake task to print also multiple lines prefixed with `|` thanks to your MultilinePatternLayout.